File: day04/day04.rb
require 'date'

data = File.readlines('input.txt').map { |x|
  [DateTime.strptime(x[1..16], "%Y-%m-%d %H:%M"), x[19..-2]] }
data.sort! { |x,y| x[0] <=> y[0] }
guards = Hash.new{ |h,k| h[k] = Array.new(60, 0) }
for d in data
  if d[1].include? "#"
    guard = d[1].match(/#(\d+)/)[1].to_i
  elsif d[1].include? "asleep"
    sleep_t = d[0].min
  else
    wake_t = d[0].min
    (sleep_t...wake_t).each { |t| guards[guard][t] += 1 }
  end
end
sleepy = guards.sort_by { |k, v| guards[k].sum }.reverse.first
p sleepy[0] * sleepy[1].index(sleepy[1].max)
sleepy2 = guards.sort_by { |k, v| guards[k].max }.reverse.first
p sleepy2[0] * sleepy2[1].index(sleepy2[1].max)

File: day04/day04.py
from datetime import datetime
import re
from collections import defaultdict

data = [(datetime.strptime(x[1:17], "%Y-%m-%d %H:%M"), x[19:-1])
         for x in open('input.txt').readlines()]
data = sorted(data, key=lambda x: x[0])
guards = defaultdict(lambda: [0 for x in range(60)])
for d in data:
    if "#" in d[1]:
        guard = int(re.search(r'#(\d+)', d[1]).group(1))
    elif "asleep" in d[1]:
        sleep = d[0].minute
    else:
        awake = d[0].minute
        for x in range(sleep, awake):
            guards[guard][x] += 1
sleepy = sorted(guards.keys(), key=lambda g: -sum(guards[g]))[0]
print(sleepy * guards[sleepy].index(max(guards[sleepy])))
sleepy2 = sorted(guards.keys(), key=lambda g: -max(guards[g]))[0]
print(sleepy2 * guards[sleepy2].index(max(guards[sleepy2])))
